### PR TITLE
RHIROS-731 Added API caching for ROS

### PR DESCRIPTION
**Pull request by patilsuraj767**
We are getting many hits on call_to_action API, reason - call_to_action API gets called from console dashboard so for every user coming to consoledot will hit call_to_action API though he/she is not a ROS user. To reduce DB connections/queries adding redis to cache the call_to_action API with TTL of 5 mins.
**Note - Do not merge this PR before merging app-interface MR linked in JIRA ticket.**

### CODE REVIEW

**Review by r14chandra — APPROVED** [review]
Looks good! Thanks @patilsuraj767 for working on it! I've tested it locally 👍🏻

**Comment by kgaikwad** [comment]
As the PROD env is stable, removed the `on Hold` label from PR.

**Comment by saltgen** [comment]
@patilsuraj767 There seems to be a NoneType error on init,

```
File "/opt/app-root/src/ros/lib/config.py", line 50, in <module>
REDIS_USERNAME = LoadedConfig.inMemoryDb.username
AttributeError: 'NoneType' object has no attribute 'username'
```

**Comment by saltgen** [comment]
@patilsuraj767 There seems to be a NoneType error on EE, is this cause of no Redis secrets?

```
File "/opt/app-root/src/ros/lib/config.py", line 50, in <module>
    REDIS_USERNAME = LoadedConfig.inMemoryDb.username
AttributeError: 'NoneType' object has no attribute 'username'
```

**Comment by kgaikwad** [comment]
> @patilsuraj767 There seems to be a NoneType error on container init,
> 
> ```
> File "/opt/app-root/src/ros/lib/config.py", line 50, in <module>
> REDIS_USERNAME = LoadedConfig.inMemoryDb.username
> AttributeError: 'NoneType' object has no attribute 'username'
> ```

@patilsuraj767 - As per explained in the [document](https://github.com/RedHatInsights/clowder/blob/master/docs/antora/modules/providers/pages/inmemorydb.adoc#clowdapp-configuration), does it require to set `inMemoryDb: true` in clowdapp file?

**Comment by patilsuraj767** [comment]
Yes, we will need `inMemoryDb: true` in `clowderapp.yaml`. Adding it.

**Review by PreetiW — COMMENTED** [review]
Thank you @patilsuraj767 for creating the PR and discussing the code details. Code wise changes looks good to me. 
When I tried to test it in the EE and tried to check the KEYS in the  engine-redis pod after hitting `call_to_action` api it was still returning empty list. I am attaching the screenshot for your reference. Can you please tell me if I am referring to correct pod here to check the redis key?

image

image

**Comment by r14chandra** [comment]
@PreetiW It's `engine` service redis pod.

**Comment by PreetiW** [comment]
> @PreetiW It's `engine` service redis pod. It should be `ros-backend-redis-*`.

Hello @r14chandra couldn't find anything like that.

**Review by r14chandra — APPROVED** [review]
@PreetiW I created an image locally as hitting the api was giving me `org_id` error (similar to the error which we were getting in local - org_id missing from `identity`->`internal`). I can see the redis pod. And also hitting the api is being cached. I think we are good. 

ros redis pod:
Screenshot from 2022-08-10 11-39-22

redis terminal output:
Screenshot from 2022-08-10 11-42-45

**Review by PreetiW — COMMENTED** [review]
Thank you @patilsuraj767 for the help with the testing, after updating the bonfire config file with local clowdapp.yaml everything started working properly in EE. LGTM 💯 

image

**Review by PreetiW — APPROVED** [review]

File: ros/api/v1/call_to_action.py [hunk 1]
```diff
@@ -4,54 +4,63 @@
...
-        system_query = system_ids_by_org_id(org_id).filter(PerformanceProfile.number_of_recommendations > 0)
-        query = (
-            db.session.query(PerformanceProfile.system_id)
-            .filter(PerformanceProfile.system_id.in_(system_query.subquery()))
-        )
-        total_system_count = query.count()
+        cache_key = "call_to_action_"+org_id
+        if res := cache.get(cache_key):
+            return res
+        else:
```
Comment (kgaikwad) [inline ros/api/v1/call_to_action.py:16]: nitpick - As you are returning result from `if`, no need of `else`.

File: ros/api/v1/call_to_action.py [hunk 2]
```diff
@@ -4,54 +4,63 @@
...
-        else:
-            if total_system_count > 1:
-                suffix = 'systems'
+            if total_system_count == 0:
+                result = {
+                    "configTryLearn": configTryLearn_Object
+                }
+                cache.set(cache_key, result)
+                return result
             else:
```
Comment (kgaikwad) [inline ros/api/v1/call_to_action.py:42]: ditto

**Review by kgaikwad — APPROVED** [review]
@patilsuraj767 - LGTM :+1:  added two inline nitpick comments.

**Comment by r14chandra** [comment]
Thanks @patilsuraj767! Merging this PR.

**Pull request merged**